internal/peer/node/unjoin.go

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package node
+
+import (
+	"github.com/hyperledger/fabric/core/ledger/kvledger"
+	"github.com/hyperledger/fabric/core/transientstore"
+	"github.com/hyperledger/fabric/internal/peer/common"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+func unjoinCmd() *cobra.Command {
+	var channelID string
+
+	cmd := &cobra.Command{
+		Use:   "unjoin",
+		Short: "Unjoin the peer from a channel.",
+		Long:  "Unjoin the peer from a channel.  When the command is executed, the peer must be offline.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if channelID == common.UndefinedParamValue {
+				return errors.New("Must supply channel ID")
+			}
+
+			if err := unjoinChannel(channelID); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+	flags := cmd.Flags()
+	flags.StringVarP(&channelID, "channelID", "c", common.UndefinedParamValue, "Channel to unjoin.")
+
+	return cmd
+}
+
+// unjoin the peer from a channel.
+func unjoinChannel(channelID string) error {
+	config := ledgerConfig()
+
+	if err := kvledger.UnjoinChannel(config, channelID); err != nil {
+		return err
+	}
+	if err := transientstore.Drop(config.RootFSPath, channelID); err != nil {
```

Comment:
This will attempt to drop the transient store at /ledgersData, however the transient store is located at /transientStore.

I know because I tried it out and received panic :-)

```
2021-07-27 17:00:23.133 EDT 0036 DEBU [transientstore] newStoreProvider -> opening provider providerPath=/Users/denyeart/work/src/github.com/hyperledger/fabric-samples/test-network-nano-bash/data/peer0.org1.example.com/ledgersData
panic: Error opening leveldb: file does not exist

goroutine 1 [running]:
github.com/hyperledger/fabric/common/ledger/util/leveldbhelper.(*DB).Open(0xc000272050)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/common/ledger/util/leveldbhelper/leveldb_helper.go:76 +0x239
github.com/hyperledger/fabric/common/ledger/util/leveldbhelper.openDBAndCheckFormat(0xc000184f20, 0x0, 0x0, 0x0)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/common/ledger/util/leveldbhelper/leveldb_provider.go:72 +0x133
github.com/hyperledger/fabric/common/ledger/util/leveldbhelper.NewProvider(0xc000184f20, 0xc000184f20, 0x5007d95, 0x10)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/common/ledger/util/leveldbhelper/leveldb_provider.go:60 +0x2f
github.com/hyperledger/fabric/core/transientstore.newStoreProvider(0xc000058d80, 0x7d, 0xc000184100, 0xc000570000, 0x88, 0x0)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/core/transientstore/store.go:115 +0x14b
github.com/hyperledger/fabric/core/transientstore.Drop(0xc000058d80, 0x7d, 0x7ffeefbff35c, 0x9, 0x0, 0x0)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/core/transientstore/store.go:287 +0x35f
github.com/hyperledger/fabric/internal/peer/node.unjoinChannel(0x7ffeefbff35c, 0x9, 0xc0002d1cd8, 0x0)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/internal/peer/node/unjoin.go:49 +0x85
github.com/hyperledger/fabric/internal/peer/node.unjoinCmd.func1(0xc000548500, 0xc00052f0a0, 0x0, 0x2, 0x0, 0x0)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/internal/peer/node/unjoin.go:29 +0xb5
github.com/spf13/cobra.(*Command).execute(0xc000548500, 0xc00052f060, 0x2, 0x2, 0xc000548500, 0xc00052f060)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/vendor/github.com/spf13/cobra/command.go:762 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0x591c3c0, 0xc0002ddf48, 0x1, 0x1)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/vendor/github.com/spf13/cobra/command.go:852 +0x2fe
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/denyeart/work/src/github.com/hyperledger/fabric/vendor/github.com/spf13/cobra/command.go:800
main.main()
	/Users/denyeart/work/src/github.com/hyperledger/fabric/cmd/peer/main.go:56 +0x4c5
```

This also highlights the need for an integration test.

Reply:
I would suggest to change the order here. Because peer determines the presence of a channel by looking at kvledger idstore entry. In the current form, there is a possibility that it leaves the data in the transient store for an non-existent channel, if it fails in between. Next time around, if user makes another attempt to delete, it may get an error that channel does not exist. In nutshell, deleting the entry from the idStore should be the last removal activity.